File: go/protocol/intents/session_test.go
package intents

import (
	"encoding/json"
	"testing"
)

func TestSessionRequestRoundTripsSharedWireFields(t *testing.T) {
	decimals := uint8(6)
	strategy := SessionPullVoucherStrategyClientVoucher
	request := SessionRequest{
		Cap:                 "1000000",
		Currency:            "USDC",
		Decimals:            &decimals,
		Network:             "devnet",
		Operator:            "operator",
		Recipient:           "recipient",
		Splits:              []SessionSplit{{Recipient: "affiliate", BPS: 250}},
		ProgramID:           "program",
		Description:         "Metered API session",
		ExternalID:          "session-001",
		MinVoucherDelta:     "1000",
		Modes:               []SessionMode{SessionModePush, SessionModePull},
		PullVoucherStrategy: &strategy,
		RecentBlockhash:     "blockhash",
	}

	if err := request.Validate(); err != nil {
		t.Fatalf("request should validate: %v", err)
	}

	encoded, err := json.Marshal(request)
	if err != nil {
		t.Fatalf("marshal failed: %v", err)
	}
	var decoded SessionRequest
	if err := json.Unmarshal(encoded, &decoded); err != nil {
		t.Fatalf("unmarshal failed: %v", err)
	}
	if decoded.ProgramID != "program" || decoded.ExternalID != "session-001" {
		t.Fatalf("decoded request lost fields: %#v", decoded)
	}
	if decoded.PullVoucherStrategy == nil || *decoded.PullVoucherStrategy != strategy {
		t.Fatalf("unexpected pull voucher strategy: %#v", decoded.PullVoucherStrategy)
	}
}

func TestSessionRequestRequiresPullStrategyWhenPullModeIsAdvertised(t *testing.T) {
	request := SessionRequest{
		Cap:       "1000000",
		Currency:  "USDC",
		Operator:  "operator",
		Recipient: "recipient",
		Modes:     []SessionMode{SessionModePull},
	}

	if err := request.Validate(); err == nil {
		t.Fatal("expected pullVoucherStrategy validation error")
	}
}

func TestOpenPayloadValidatesModeSpecificFields(t *testing.T) {
	push := OpenPayload{
		Action:           "open",
		Mode:             SessionModePush,
		ChannelID:        "channel",
		Deposit:          "1000000",
		AuthorizedSigner: "session-signer",
		Signature:        "signature",
	}
	if err := push.Validate(); err != nil {
		t.Fatalf("push open should validate: %v", err)
	}

	pull := OpenPayload{
		Action:           "open",
		Mode:             SessionModePull,
		TokenAccount:     "token-account",
		ApprovedAmount:   "1000000",
		AuthorizedSigner: "session-signer",
		Signature:        "signature",
	}
	if err := pull.Validate(); err != nil {
		t.Fatalf("pull open should validate: %v", err)
	}
}

func TestSignedVoucherAcceptsCumulativeAlias(t *testing.T) {
	var voucher SignedVoucher
	err := json.Unmarshal([]byte(`{
		"data": {
			"channelId": "channel",
			"cumulative": "25000",
			"expiresAt": 4102444800
		},
		"signature": "signature"
	}`), &voucher)
	if err != nil {
		t.Fatalf("unmarshal failed: %v", err)
	}
	if err := voucher.Validate(); err != nil {
		t.Fatalf("voucher should validate: %v", err)
	}
	if voucher.Data.CumulativeAmount != "25000" {
		t.Fatalf("expected cumulative alias, got %q", voucher.Data.CumulativeAmount)
	}
}

func TestMeteringDirectiveAndCommitReceiptValidateWireFields(t *testing.T) {
	directive := MeteringDirective{
		DeliveryID: "delivery-001",
		SessionID:  "channel",
		Amount:     "5000",
		Currency:   "USDC",
		Sequence:   1,
		ExpiresAt:  DefaultSessionExpiresAt,
		CommitURL:  "https://merchant.example/session/commit",
		Proof:      "proof",
	}
	if err := directive.Validate(); err != nil {
		t.Fatalf("directive should validate: %v", err)
	}

	receipt := CommitReceipt{
		DeliveryID: "delivery-001",
		SessionID:  "channel",
		Amount:     "5000",
		Cumulative: "30000",
		Status:     CommitStatusCommitted,
	}
	if err := receipt.Validate(); err != nil {
		t.Fatalf("receipt should validate: %v", err)
	}
}

func TestCommitReceiptRejectsUnknownStatus(t *testing.T) {
	receipt := CommitReceipt{
		DeliveryID: "delivery-001",
		SessionID:  "channel",
		Amount:     "5000",
		Cumulative: "30000",
		Status:     CommitStatus("accepted"),
	}

	if err := receipt.Validate(); err == nil {
		t.Fatal("expected status validation error")
	}
}

File: php/tests/SessionTest.php
<?php

declare(strict_types=1);

namespace SolanaMpp\Tests;

use InvalidArgumentException;
use PHPUnit\Framework\TestCase;
use SolanaMpp\Intent\SessionCommitReceipt;
use SolanaMpp\Intent\SessionMeteringDirective;
use SolanaMpp\Intent\SessionRequest;
use SolanaMpp\Intent\SessionSplit;
use SolanaMpp\Intent\SignedVoucher;
use SolanaMpp\Intent\VoucherData;

final class SessionTest extends TestCase
{
    public function testSessionRequestSerializesSharedWireFields(): void
    {
        $request = new SessionRequest(
            cap: '1000000',
            currency: 'USDC',
            operator: 'operator',
            recipient: 'recipient',
            decimals: 6,
            network: 'devnet',
            splits: [new SessionSplit(recipient: 'affiliate', bps: 250)],
            programId: 'program',
            description: 'Metered API session',
            externalId: 'session-001',
            minVoucherDelta: '1000',
            modes: [SessionRequest::MODE_PUSH, SessionRequest::MODE_PULL],
            pullVoucherStrategy: SessionRequest::PULL_CLIENT_VOUCHER,
            recentBlockhash: 'blockhash',
        );

        self::assertSame([
            'cap' => '1000000',
            'currency' => 'USDC',
            'operator' => 'operator',
            'recipient' => 'recipient',
            'decimals' => 6,
            'network' => 'devnet',
            'splits' => [['recipient' => 'affiliate', 'bps' => 250]],
            'programId' => 'program',
            'description' => 'Metered API session',
            'externalId' => 'session-001',
            'minVoucherDelta' => '1000',
            'modes' => ['push', 'pull'],
            'pullVoucherStrategy' => 'clientVoucher',
            'recentBlockhash' => 'blockhash',
        ], $request->toArray());
    }

    public function testSessionRequestRequiresPullVoucherStrategyForPullMode(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('pullVoucherStrategy is required');

        new SessionRequest(
            cap: '1000',
            currency: 'USDC',
            operator: 'operator',
            recipient: 'recipient',
            modes: [SessionRequest::MODE_PULL],
        );
    }

    public function testSignedVoucherSerializesCumulativeVoucher(): void
    {
        $voucher = new SignedVoucher(
            data: new VoucherData(
                channelId: 'channel',
                cumulativeAmount: '25000',
                expiresAt: VoucherData::DEFAULT_EXPIRES_AT,
                nonce: 1,
            ),
            signature: 'signature',
        );

        self::assertSame([
            'data' => [
                'channelId' => 'channel',
                'cumulativeAmount' => '25000',
                'expiresAt' => 4_102_444_800,
                'nonce' => 1,
            ],
            'signature' => 'signature',
        ], $voucher->toArray());
    }

    public function testMeteringDirectiveSerializesCommitFields(): void
    {
        $directive = new SessionMeteringDirective(
            deliveryId: 'delivery-001',
            sessionId: 'channel',
            amount: '5000',
            currency: 'USDC',
            sequence: 1,
            expiresAt: VoucherData::DEFAULT_EXPIRES_AT,
            commitUrl: 'https://merchant.example/session/commit',
            proof: 'proof',
        );

        self::assertSame('delivery-001', $directive->toArray()['deliveryId']);
        self::assertSame('channel', $directive->toArray()['sessionId']);
        self::assertSame('https://merchant.example/session/commit', $directive->toArray()['commitUrl']);
    }

    public function testCommitReceiptRejectsUnknownStatus(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('status must be committed or replayed');

        new SessionCommitReceipt(
            deliveryId: 'delivery-001',
            sessionId: 'channel',
            amount: '5000',
            cumulative: '30000',
            status: 'accepted',
        );
    }
}
